src/libfetchers/tarball.cc

```diff
@@ -35,9 +35,12 @@ DownloadFileResult downloadFile(
         };
     };
 
-    if (cached && !cached->expired)
+    if (cached && (!cached->expired || settings.offline))
         return useCached();
```

Comment:
Maybe log something (I’d say `warn`, but that might be too noisy) if we keep the expired cache because we’re offline

Reply:
I think there are different intuitions on how Nix caching should work, but a message here is worthwhile for debugging.